### ISSUE TITLE
Does using the word "KMS" confer any advantage to this project?

### ISSUE THREAD

**Issue by jMyles** [nucypher/nucypher-kms#263]
Since the beginning of this project, we have called it `nucypher-kms`.  But does the "KMS" part of the name add anything useful?  Are we in fact building a "KMS"?  Is it even possible, given the current use of this word, to determine with clarity exactly what a "KMS" is?

----

As this project has progressed, I think it has become clear that what we're building isn't primarily a "Key Management System."

First, let's consider exactly what a "Key Management System" is, and whether this phrase is unambiguous enough for us to use as part of the name of this project (hint: I think it isn't).  Let's also consider whether this phrase is sufficiently widespread to give this project any name-recognition advantage by using it.

### What is a KMS?

The [Wikipedia disambiguation page for "KMS"](https://en.wikipedia.org/wiki/KMS) lists, under the "Technology" heading:
```
    Kernel mode-setting, setting up screen mode and resolution in operating system kernels
        KMS driver, Linux kernel device driver for display controllers
    KMS (hypertext), the Knowledge Management System hypertext system
    Kaspersky Mobile Security
    Key Management Service, a Microsoft technology
    Kerio MailServer
    Kwangmyŏngsŏng (disambiguation), a class of experimental satellites developed by North Korea
```
None of these have anything to do with cryptographic tooling.  Even the MS "Key Management Service" is not about cryptographic keys - it's about Microsoft license keys. 

*In other words, there is literally nothing on the Wikipedia page for the word "KMS" that has anything to do with this project.*  I think that says something.  I think this alone is a strong indication that using the word "KMS" will create confusion rather than relieving it.

But let's keep reading to see if "KMS" makes sense when examining its actual uses.  Hopping over to the [Wikipedia page for "Key Management"](https://en.wikipedia.org/wiki/Key_management).  Here, we find, under a subheading, "Key management system":

> A key management system (KMS), also known as a cryptographic key management system (CKMS), is an integrated approach for generating, distributing and managing cryptographic keys for devices and applications. They may cover all aspects of security - from the secure generation of keys over the secure exchange of keys up to secure key handling and storage on the client. Thus, a KMS includes the backend functionality for key generation, distribution, and replacement as well as the client functionality for injecting keys, storing and managing keys on devices.

So this description is closer obviously.  It does cover some of the basic mechanics that drive this project.  But does it actually describe anything user-facing?

This project does indeed feature "integrated approach for generating, distributing and managing cryptographic keys."  But is that its primary purpose?  Is that what makes it different or useful?  Neither Alice, Bob, nor Ursula ever even become "aware" that keys exist.  They never need to handle keys to *do* anything with them.  

The user here has far more interesting interfaces to work with than "Keys" - they get fully built nodes, Policies, DataSources, and MessageKits.  Rather than a crude "KMS", what we have here is a feature-rich network for doing types of communication that aren't even possible in the context of a "KMS" alone.

Nor does this description describe features like the payment / medallion model, DisappearingAlice, or the capsule / label system.  These are the things that make this project different and interesting, and none of them are captured by this (or any) description of a "KMS".

OK, but maybe WikiPedia just isn't a good source of textual definition for this concept.  Let's instead look at how this phrase is used in the wild.

### Is this thing similar to AWS KMS?

Googling "Key management system" brings, as the first result, AWS KMS.  Now, first, note something important here: Amazon uses the phrase "Key Management Service (KMS)," not "Key Management System."  So it seems immediately that there's a disparity about what "KMS" even means: is it a *system* or a *service*?

Amazon's description of their Key Management Serivce: 

> AWS Key Management Service (KMS) is a managed service that makes it easy for you to create and control the encryption keys used to encrypt your data, and uses FIPS 140-2 validated hardware security modules to protect the security of your keys.

Is that what you'd say about this project?  I don't think so.  If I had to write a description of this project in exactly the same tembre as this one, I'd say something like:

> NuCypher is a decentralized service that makes is possible for you to define recipients for private data and then to build (or encourage others to build) applications that share data with them.  To do this, it uses a distributed swarm of nodes which perform proxy re-encryption, and which never have access to your private data.

### What about other KMS's?

Back to Google - the [second result is the Microsoft Technet article on "Understanding KMS."](https://technet.microsoft.com/en-us/library/ff793434.aspx#E1B)  It reads:

> KMS activates computers on a local network, eliminating the need for individual computers to connect to Microsoft. To do this, KMS uses a client–server topology. 

In other words, the *second result on Google for "KMS"* is about a *completely* unrelated concept.  I'm left with the feeling that using the phrase "KMS" *adds* confusion to our project rather than subtracting it.

### What do we say instead?

So, in a nutshell, what I'm trying to say is:

**If we give the impression that NuCypher is a "KMS", it might not be clear what that even means.  In the best-case scenario, we hope that it evokes an image of a drop-in replacement for things like Amazon KMS, but even in this case, we're both over- and underselling it.  This comparison misses the *main things* that make NuCypher different and compelling, and fails to encourage people to imagine the best use cases.**

Furthermore, "KMS" is not a phrase that is of such household-name status and disambiguity as to gain us anything interesting by trying to leverage it.  

**So, I propose that we call this thing simply "NuCypher."**

Now, I understand that the word "NuCypher" is currently a name given to the organization, and that under its umbrella lie multiple products with different use-cases.  But now might be time to change that, just as we have done with our website.  The website certainly gives the impression that this project is the entire kit and caboodle.  In fact, our website features the word "KMS" only once: 

> NuCypher KMS helps dApp developers secure their data on the blockchain with decentralized proxy re-encryption as a service.

And of course, this reads equally well if the word "KMS" is simply removed:

> NuCypher helps dApp developers secure their data on the blockchain with decentralized proxy re-encryption as a service.

**Comment by cygnusv** [comment]
I agree with you to a certain extent. Indeed, "KMS" doesn't seem the most accurate name for this project (at least in its current state). If anything, this is more a "descentralized data sharing service" (or similar). Moreover, to the layman, "KMS" isn't something descriptive. In fact, I bet that most people interested in the project simply say "NuCypher" to refer to it, perhaps because the limitations of "KMS".

On the other hand, I think changing the name at this stage may make the project look a little bit erratic. With respect to the changing the name "NuCypher KMS" to just "NuCypher", I would certainly prefer keeping NuCypher for the company, and "NuCypher _<insert name here>_" for products/services/etc. In that case, I'd propose something in the like of "NuCypher Network" (or similar).

**Comment by jMyles** [comment]
> If anything, this is more a "descentralized data sharing service" (or similar). Moreover, to the layman, "KMS" isn't something descriptive. 
> On the other hand, I think changing the name at this stage may make the project look a little bit erratic.

I don't think so - it's not like "NuCypher KMS" is a widely used name anyway.  I think people will barely notice if we just drop the "KMS".  Looking around the internet, most people seem to just be calling it "NuCypher" already.  Most the references to the three letters "KMS" are from the token symbol we briefly considered, "NKMS".  But if we're going to go with a different symbol, that doesn't seem to matter anyway.

> In that case, I'd propose something in the like of "NuCypher Network" (or similar).

Yeah, I mean, I'm fine with "NuCypher Network."  But even if we just call it "NuCypher", that doesn't stop us from naming future projects "NuCypher <thing>".

**Comment by tuxxy** [comment]
> Since the beginning of this project, we have called it nucypher-kms. But does the "KMS" part of the name add anything useful? Are we in fact building a "KMS"? Is it even possible, given the current use of this word, to determine with clarity exactly what a "KMS" is?

Yes, we are building a KMS. As you found in your post, a KMS is a key management system. In my opinion, this is what NuCypher is. It is the first of its kind -- a decentralized, consumer (and enterprise) grade KMS. It's a decentralized answer to centralized key management tools. Our blog post from July 2017 is one of the first results when you search "Decentralized Key Management" on Google. In fact, we are the majority of the first page of Google search results. I'm quite confident in our users to grasp the idea of what a KMS is.

> The Wikipedia disambiguation page for "KMS" lists, under the "Technology" heading:

In this list I believe it is a mistake that "Key Management System" is not placed under this. In fact, if you navigate back to that page, you will see that I have added "Key Management System" to this list. I've even added "NuCypher" to the list of KMSs for the respective Wikipedia page. I don't think KMS is an unknown word. It's quite well known in the developer and security worlds for being used to handle user and data encryption keys where necessary. For many AWS and Google Cloud developers, they intrinsically rely upon the platform and _cannot switch_ because they cannot lose this KMS functionality.

Onto your description:
> So this description is closer obviously. It does cover some of the basic mechanics that drive this project. But does it actually describe anything user-facing?

No, not quite. Historically, KMSs have _only_ been enterprise grade and have never faced the user before. To do so would have intrinsically introduced security issues as they are centralized services leaving liability for user mistakes to the service provider.

This uncertainty you are expressing is the effect of being a groundbreaker in this space. We are the _first consumer_ grade KMS. This is needed because it must be used in decentralized applications. Decentralization radically shifts the power in the hands of the consumer; anything less would leave us with centralized services.

> Neither Alice, Bob, nor Ursula ever even become "aware" that keys exist. They never need to handle keys to do anything with them.

This is objectively false. For Alice to encrypt data, they must get a symmetric key from Umbral. For Bob to decrypt data, they must use a symmetric key from Umbral. For Ursula to re-encrypt data, they must use a re-encryption key from Umbral. All of our characters in this narrative _must_ handle keys. Just because we do not expose this to them in our high-level API doesn't negate this fact -- it _still_ happens. Honestly, this speaks to the _management_ part of the KMS for me and tells me we have done it very well.

> The user here has far more interesting interfaces to work with than "Keys" - they get fully built nodes, Policies, DataSources, and MessageKits. Rather than a crude "KMS", what we have here is a feature-rich network for doing types of communication that aren't even possible in the context of a "KMS" alone.

This is only true if you are Alice building out a large integration into your application. This is not true for Ursula who will simply run a binary, nor is it true for Bob who only has to make a threshold number of re-encryption requests via REST.

> Nor does this description describe features like the payment / medallion model, DisappearingAlice, or the capsule / label system. These are the things that make this project different and interesting, and none of them are captured by this (or any) description of a "KMS".

Again, this is simply because we are building cutting edge technology. The KMS space hasn't had the opportunity to be disrupted this much. Proxy Re-encryption is relatively new in cryptography years. Elliptic Curve Cryptography wasn't introduced until 1985, and then it took 20 years before it was actually adopted. Then it took a few more years after that to become standard.

Proxy Re-encryption, on the other hand, was first given a serious answer in 1998 and it didn't offer any key privacy -- a large adoption problem -- nor was it unidirectional. In fact, formal security models and definitions weren't until 2005 by our own Giuseppe Anteniese _et al_ and Canetti _et al_ (in 2007). The latter defined _IND-PRE-CCA_ security.

Obviously, the entire tech world is relatively young to similar businesses. I'm sure you know how hard it is to innovate tools when nearly the entire world relies on one variant. This definitely rings true for key management systems. There simply hasn't been enough work on them to produce such an incredible thing as we have. Not to mention, they all must remain compliant for certain companies to use them -- this is not a constraint in the dApp, blockchain, or consumer space.

> Amazon uses the phrase "Key Management Service (KMS)," not "Key Management System." So it seems immediately that there's a disparity about what "KMS" even means: is it a system or a service?

Quite simply, AWS is a collection of _services_. They are largely _<blank> as a Service_. They are following the SaaS model with this name and I don't think it should be included as an argument against system vs service. If anything, AWS KMS is simply a service that offers access to the AWS Key Management _System_.

> In other words, the second result on Google for "KMS" is about a completely unrelated concept. I'm left with the feeling that using the phrase "KMS" adds confusion to our project rather than subtracting it.

I don't think SEOd results are valid arguments against _what_ something is intrinsically. This is quite stretching, imho.

> Furthermore, "KMS" is not a phrase that is of such household-name status and disambiguity as to gain us anything interesting by trying to leverage it.

I think this is a plus that plays in our favor. Every blockchain company, every dApp developer, before or after we launch mainnet, will _know_ what a KMS is _and_ they will think NuCypher when they do.

> So, I propose that we call this thing simply "NuCypher."

This is what I've always called it. I have never presented it as "NuCypher KMS is a <blank>". I have always presented it as "NuCypher is a <blank>". I think our developer slides display this well enough and our presentations also document this fact.

tl;dr: NuCypher _is_ a KMS. It's just built for a brand new paradigm and a brand new market, as such the lines are greyed. NuCypher _will_ become a household name in this market and space. Bringing the idea of a KMS to a community that desperately _needs_ secure software is one step to making this happen.

**Comment by jMyles** [comment]
> Yes, we are building a KMS. As you found in your post, a KMS is a key management system.

I mean, no - "key management system" is not one of the most popular phrases paired with the abbreviation "KMS".  In fact, neither Amazon nor Google call their KMS that - they say "key management service" (also not a popular phrase paired with this abbreviation).

> In this list I believe it is a mistake that "Key Management System" is not placed under this. In fact, if you navigate back to that page, you will see that I have added "Key Management System" to this list.

OK, well that's good.  I think you did a good deed.  But is this the mission of our organization?  To change the world's mind on what a "KMS" is?

> I don't think KMS is an unknown word. It's quite well known in the developer and security worlds for being used to handle user and data encryption keys where necessary.

I disagree.  I had used Amazon KMS before being invovled at NuCypher and I still wasn't sure what "KMS" in "NuCypher KMS" meant.

> This uncertainty you are expressing is the effect of being a groundbreaker in this space. We are the first consumer grade KMS.

Yeah, I figured you were going to make this argument, and it's a good one.  Whatever we are, we are the first consumer-facing thing like it.  Fine.  But does that mean that we're primarily in the business of *managing keys*?  Is that the most important thing to communicate?

> Just because we do not expose this to them in our high-level API doesn't negate this fact -- it still happens.

I need to generate and use symmetric keys for DH in every HTTPS negotiation.  Is my browser a KMS?  I need to store and use keys with SSH.  Is openssh a KMS?  Is Metamask a KMS?  Is Signal a KMS?  On and on and on.  Just because something uses cryptographic keys somewhere under the hood does not mean that its primary concern is managing keys.  NuCypher isn't primarily about managing keys, it's about providing a privacy layer via Policies and DataSources.  Unlike Amazon KMS and Google KMS, the user can't even effectively reason about a key.  They aren't managing or using keys except with the same level of indirection as Firefox, openssh, Metamask, or Signal.

> This is only true if you are Alice building out a large integration into your application. This is not true for Ursula who will simply run a binary, nor is it true for Bob who only has to make a threshold number of re-encryption requests via REST.

I don't really understand what you're saying here.  All three Characters do things that they don't do in any other KMS and that have nothing whatsoever to do with key management.

> Again, this is simply because we are building cutting edge technology. The KMS space hasn't had the opportunity to be disrupted this much.

I want to suggest that there *is no* consistent, coherently defined KMS-space, and for good reason: keys aren't something users care about.  Ever.  They're a weak metaphor for a mathematical operation, and it's the ability to leverage that operation in applications that users want.  It's possible that nobody will use the word "KMS" to describe things like Amazon KMS or Google KMS in 10 years.  It's possible that people will only use it to describe the (completely unrelated) Microsoft KMS.  Or that they won't use it for either of these things.

I don't think we're well-positioned to put effort into "disrupting the KMS space."  The world doesn't need KMS's, it needs coherent models of privacy.  Ideally, this is done without "keys" at all, and a more sober metaphor replaces it.

> This definitely rings true for key management systems. There simply hasn't been enough work on them to produce such an incredible thing as we have.

Well, then why *name* our thing after them at all?  If we're the best in the business, can't we pick a more accurate name?

> Quite simply, AWS is a collection of services. They are largely as a Service. They are following the SaaS model with this name and I don't think it should be included as an argument against system vs service. If anything, AWS KMS is simply a service that offers access to the AWS Key Management System.

Sure - that was my impression too.  And it's a great indictment against any kind of universality of the word "KMS" - if Amazon or Google can, at their leisure, change "System" to "Service" without anyone complaining that they've reappropriated the term, then the term is *weak*.

> I don't think SEOd results are valid arguments against what something is intrinsically. This is quite stretching, imho.

This isn't an argument about what something is intrinsically.  This is an argument about what is the best word or phrase to describe our project.  Given that a search for one of the words in our name turns up all kinds of crazy unrelated results, I think that's pretty damn compelling that we picked an incorrect word.

> I think this is a plus that plays in our favor. Every blockchain company, every dApp developer, before or after we launch mainnet, will know what a KMS is and they will think NuCypher when they do.

...but why are we trying to push "KMS" as a brand in the first place then?  I don't see any advantage to this project to "every blockchain company, every dApp developer, before or after we launch mainnet, [knowing] what a KMS is", and I'm not particularly excited about putting in effort for this cause.  By dropping "KMS" from the name, we ensure that they all instead just know *NuCypher*.

> This is what I've always called it. I have never presented it as "NuCypher KMS is a ". I have always presented it as "NuCypher is a ". I think our developer slides display this well enough and our presentations also document this fact.

If this is what you have always called it, then are you agreeable to changing the name (in the repo, documentation, and website)?

----

> tl;dr: NuCypher is a KMS.

You haven't actually advanced this point at all.

In what ways is NuCypher a KMS?  Does it provide access to keys so that users can use them in their own applications?  No.  Does it help users manage keys that already exist?  No.  Exactly what kind of "key management" is happening here that doesn't happen in *any* application that uses encryption?

If you show this project to people, with no name at all, do you think anybody will say, "oh, it's a KMS!"  Of course not.  There's no actual key management happening.

And again, conversely, the phrase "KMS", for the extremely limited audience who will realize that we're making a reference to things like Amazon KMS, hugely limits the perception of what NuCypher does.  If I fork Firefox and change every instance of "Web Browser" to "Diffie-Hellman Key Exchange System" (which of course, Firefox does), don't you think I'm pointing people in the wrong place to look for the killer features?  That's exactly what calling this a "KMS" does - it points them to the fact that this will manage keys for them instead of the fact that it will allow them to create Policies and DataSources and MessageKits.  

Now, I think you're on to something when you make the point that people just aren't used to seeing key management used to do something so cool.  On that point, we absolutely agree.  And for that reason, I think that we gain nothing by invoking the idea that NuCypher is like Amazon KMS.

If you want to keep documentation that says, "NuCypher can be used instead of Amazon KMS to achieve safe handling of sensitive key material on both servers and clients, etc." - I think that's great.  A vanilla KMS is one potential application for NuCypher.  But it is not what most users will care about, and it's not what will launch this code into world-changing projects.

**Comment by tuxxy** [comment]
> I mean, no - "key management system" is not one of the most popular phrases paired with the abbreviation "KMS". In fact, neither Amazon nor Google call their KMS that - they say "key management service" (also not a popular phrase paired with this abbreviation).

In this point and the ones below it, we are disagreeing on the nature of services that allow access to systems. I am not going to dive deeper in this topic, but it's clear that Amazon provides services; thus the word "service" in the name. It's quite clear to me that their service is providing access to a system. I don't think we're going to ever agree on this, so let's move beyond this.

> OK, well that's good. I think you did a good deed. But is this the mission of our organization? To change the world's mind on what a "KMS" is?

No, but a major mission (at least in my opinion) is to give and introduce dApp developers to a DKMS so that they can build tools that enable privacy; thus, the idea of a privacy layer. 

> But does that mean that we're primarily in the business of managing keys? Is that the most important thing to communicate?

What else do we manage? Fragments? Those are certainly _key_ fragements. We certainly don't manage data.

> I need to generate and use symmetric keys for DH in every HTTPS negotiation. Is my browser a KMS?

Clearly not, but you're disregarding a major component of managing keys -- revocation. In a KMS, you can revoke access to a certain key as needed. Similarly, your browser doesn't allow you to arbitrarily store keys generated for other users, either.

> I don't really understand what you're saying here. All three Characters do things that they don't do in any other KMS and that have nothing whatsoever to do with key management.

Generating, storing, re-encrypting, decrypting, revoking, etc aren't functions of management to you?

> Given that a search for one of the words in our name turns up all kinds of crazy unrelated results, I think that's pretty damn compelling that we picked an incorrect word.

Try searching for "decentralized key management system".

> ...but why are we trying to push "KMS" as a brand in the first place then?

We're not, we're pushing NuCypher as the brand. The KMS is just one of many of the future tools NuCypher intends on creating. This application, to me, will forever be called "The KMS".

> If this is what you have always called it, then are you agreeable to changing the name (in the repo, documentation, and website)?

Maybe, but I think we have more important issues on our hand now (i.e: getting to testnet/mainnet). If any changes occur, they should happen after test or after mainnet.

> You haven't actually advanced this point at all.

I disagree.

> Does it provide access to keys so that users can use them in their own applications? No. Does it help users manage keys that already exist? No. Exactly what kind of "key management" is happening here that doesn't happen in any application that uses encryption?

Your first two questions -- we actually do, yes. Your last question is obvious. Think about all the management functions we perform.

> There's no actual key management happening.

I completely disagree, but it seems I will not be able to convince you. If you ask a NuCypher user "who manages your keys", they would say "NuCypher".

**Comment by michwill** [comment]
I think, developers interface with NuCypher as with a KMS. In this sense, when communicating to developers, we can call it a KMS. Although, there are differences.

But when we want to communicate use cases, they seem to way broader than a typical KMS indeed. Maybe, the name can derive from the primary function - managing permissions to work with data while it's encrypted

**Comment by jMyles** [comment]
> I think, developers interface with NuCypher as with a KMS. In this sense, when communicating to developers, we can call it a KMS. Although, there are differences.

Yeah?  I mean, every single one of Amazon KMS's APIs use a *key* object.  NuCypher is totally different in this regard - we never once mention a "key" to the user except in our current (and hopefully soon to be changed) name.

I don't think developers interface in the spirit of a KMS - they get a much higher level management system and never even have to consider keys.  

(in the intervening moments here, you commented on Discord: "Ok, maybe we should go a level higher.  We are managing permissions to work with data.")

> Maybe, the name can derive from the primary function - managing permissions to work with data while it's encrypted

Well, I'm not too interested in inventing a new name at this late stage.  "NuCypher" is already a pretty cool name and has gained a ton of traction.  I think people who know about NuCypher are already totally prepared to use (and in some cases are already using) the word "NuCypher" in precisely the same way that we internally sometimes use the phrase "the KMS."
